tests/cases/upsample.py

```diff
@@ -87,19 +87,94 @@ def test_output(self):
                 self.assertTrue(np.array_equal(array.data, data), str(array_key))
 
             elif array_key == ArrayKeys.RAW_UPSAMPLED:
-
-                self.assertTrue(array.data[0,0,0] == 30)
-                self.assertTrue(array.data[1,0,0] == 30)
-                self.assertTrue(array.data[2,0,0] == 31)
-                self.assertTrue(array.data[3,0,0] == 31)
+                self.assertEqual(array.data[0,0,0], 30)
+                self.assertEqual(array.data[1,0,0], 30)
+                self.assertEqual(array.data[2,0,0], 31)
+                self.assertEqual(array.data[3,0,0], 31)
 
             elif array_key == ArrayKeys.GT_LABELS_UPSAMPLED:
 
-                self.assertTrue(array.data[0,0,0] == 0)
-                self.assertTrue(array.data[1,0,0] == 0)
-                self.assertTrue(array.data[2,0,0] == 1)
-                self.assertTrue(array.data[3,0,0] == 1)
+                self.assertEqual(array.data[0,0,0], 0)
+                self.assertEqual(array.data[1,0,0], 0)
+                self.assertEqual(array.data[2,0,0], 1)
+                self.assertEqual(array.data[3,0,0], 1)
 
             else:
 
                 self.assertTrue(False, "unexpected array type")
+
+    def test_crop(self):
+
+        ArrayKey('RAW_UPSAMPLED')
+        ArrayKey('GT_LABELS_UPSAMPLED')
+
+        request = BatchRequest()
+        request.add(ArrayKeys.RAW_UPSAMPLED, (116,120,120))
+        request.add(ArrayKeys.GT_LABELS, (200,200,200))
+        request.add(ArrayKeys.GT_LABELS_UPSAMPLED, (196,196,196))
+
+        pipeline = (
+                UpSampleTestSource() +
+                UpSample(ArrayKeys.RAW, 2, ArrayKeys.RAW_UPSAMPLED) +
+                UpSample(ArrayKeys.GT_LABELS, 2, ArrayKeys.GT_LABELS_UPSAMPLED)
+        )
+
+        with build(pipeline):
+            batch = pipeline.request_batch(request)
+
+        for (array_key, array) in batch.arrays.items():
+            for dim, data_length in enumerate(array.data.shape):
+                roi_dim = array.spec.roi.get_shape()[dim]
+                voxel_size_dim = array.spec.voxel_size[dim]
+                self.assertEqual(data_length, roi_dim / voxel_size_dim)
+            # assert that pixels encode their position for supposedly unaltered
+            # arrays
+            if array_key in [ArrayKeys.RAW, ArrayKeys.GT_LABELS]:
+                continue
```

Comment:
A large block of unreachable code here, can it all be deleted?
It looks like this is basically just doing the same thing as the `provide` method of the test source, so I guess it just tests that the `np.meshgrid` function doesn't change. I'd say it should be removed from the other tests as well.